### PR TITLE
Unify Common Test Fixture Properties (Note Possible API Changes)

### PR DESCRIPTION
**Pull request by DavidBakerEffendi**
Two main groups of properties are commonly duplicated among test fixtures:
* Dataflow processing with custom semantic definitions
* Post processing passes

This PR uses a combination of traits and new additions to `TestCpg` to improve the re-usability and consistency of these among frontend fixtures.

I leave it up to the authors of the frontends to refactor these changes in each test suite.themselves. The result is that all test fixtures include the options for
```scala
.withOssDataflow(Boolean)
.withExtraFlows(List[FlowSemantic])
.withPostProcessingPasses(Boolean)
```

Misc: Added a fix for #4019

### CODE REVIEW

File: joern-cli/frontends/x2cpg/src/test/scala/io/joern/x2cpg/testfixtures/TestCpg.scala [hunk 1]
```diff
@@ -26,6 +27,12 @@ abstract class TestCpg extends Cpg() with LanguageFrontend {
 
   protected def applyPasses(): Unit
 
+  private def applyPostProcessingPassesIfEnabled(): Unit = if (!_withPostProcessing) {
```
Comment (xavierpinho) [inline joern-cli/frontends/x2cpg/src/test/scala/io/joern/x2cpg/testfixtures/TestCpg.scala]: I might be reading this wrong, but is the negated condition what you intended?

**Review by xavierpinho — APPROVED** [review]

File: joern-cli/frontends/x2cpg/src/test/scala/io/joern/x2cpg/testfixtures/TestCpg.scala (hunk 1 above)
Comment (DavidBakerEffendi) [reply joern-cli/frontends/x2cpg/src/test/scala/io/joern/x2cpg/testfixtures/TestCpg.scala]: Indeed this was the cause of the bugs that cropped up :D

File: dataflowengineoss/build.sbt [hunk 2]
```diff
@@ -1,6 +1,6 @@
 name := "dataflowengineoss"
 
-dependsOn(Projects.semanticcpg, Projects.x2cpg)
+dependsOn(Projects.semanticcpg, Projects.x2cpg % "compile->compile;test->test")
```
Comment (mpollmeier) [inline dataflowengineoss/build.sbt:3]:
```suggestion
dependsOn(Projects.semanticcpg, Projects.x2cpg, Projects.x2cpg % "test->test")
```

The `;` separator led to problems with intellij in the past - splitting them up was the fix back then. Not sure if the problem still exists, but better play it safe. 

Just noticed that this occurs in other places again - I'll send a PR for that after this is merged.

File: joern-cli/frontends/c2cpg/build.sbt [hunk 3]
```diff
@@ -1,6 +1,6 @@
 name := "c2cpg"
 
-dependsOn(Projects.semanticcpg, Projects.dataflowengineoss % Test, Projects.x2cpg % "compile->compile;test->test")
+dependsOn(Projects.semanticcpg, Projects.dataflowengineoss % "compile->compile;test->test", Projects.x2cpg % "compile->compile;test->test")
```
Comment (mpollmeier) [inline joern-cli/frontends/c2cpg/build.sbt:3]: same here, but feel free to keep as is, I'll clean this up after this PR is in

File: joern-cli/frontends/javasrc2cpg/src/test/scala/io/joern/javasrc2cpg/testfixtures/JavaSrcCodeToCpgFixture.scala [hunk 4]
```diff
@@ -23,31 +22,25 @@ trait JavaSrcFrontend extends LanguageFrontend {
...
 }
 
-class JavaSrcCode2CpgFixture(withOssDataflow: Boolean = false, enableTypeRecovery: Boolean = false)
-    extends Code2CpgFixture(() => new JavaSrcTestCpg(enableTypeRecovery).withOssDataflow(withOssDataflow)) {
+class JavaSrcCode2CpgFixture(
+  withOssDataflow: Boolean = false,
+  extraFlows: List[FlowSemantic] = List.empty,
+  enableTypeRecovery: Boolean = false
+) extends Code2CpgFixture(() =>
+      new JavaSrcTestCpg(enableTypeRecovery).withOssDataflow(withOssDataflow).withExtraFlows(extraFlows)
```
Comment (mpollmeier) [inline joern-cli/frontends/javasrc2cpg/src/test/scala/io/joern/javasrc2cpg/testfixtures/JavaSrcCodeToCpgFixture.scala:39]: interesting formatting... did scalafmt come up with that? in that case we need to update our config i guess... :)

File: joern-cli/frontends/rubysrc2cpg/src/test/scala/io/joern/rubysrc2cpg/testfixtures/RubyCode2CpgFixture.scala [hunk 5]
```diff
@@ -29,46 +29,37 @@ trait RubyFrontend(useDeprecatedFrontend: Boolean) extends LanguageFrontend {
...
+  extraFlows: List[FlowSemantic] = List.empty,
   packageTable: Option[PackageTable] = None,
   useDeprecatedFrontend: Boolean = false
 ) extends Code2CpgFixture(() =>
-      new DefaultTestCpgWithRuby(withPostProcessing, withDataFlow, packageTable, useDeprecatedFrontend)
-    ) {
-
-  implicit val resolver: ICallResolver           = NoResolve
-  implicit lazy val engineContext: EngineContext = EngineContext()
+      new DefaultTestCpgWithRuby(packageTable, useDeprecatedFrontend)
```
Comment (mpollmeier) [inline joern-cli/frontends/rubysrc2cpg/src/test/scala/io/joern/rubysrc2cpg/testfixtures/RubyCode2CpgFixture.scala:55]: again, strange indentation, no?

**Review by mpollmeier — APPROVED** [review]

File: joern-cli/frontends/rubysrc2cpg/src/test/scala/io/joern/rubysrc2cpg/testfixtures/RubyCode2CpgFixture.scala (hunk 5 above)
Comment (DavidBakerEffendi) [reply joern-cli/frontends/rubysrc2cpg/src/test/scala/io/joern/rubysrc2cpg/testfixtures/RubyCode2CpgFixture.scala:55]: Yeah it is pretty strange agreed

**Comment by DavidBakerEffendi** [comment]
It seems in some frontends, post-processing and data-flow are not commutative, and thus the order of operations need to be specific. The result is that `SemanticTestCpg` does not override `applyPasses` anymore and simply exposes `applyOssDataflow`

**Pull request merged**